### ISSUE TITLE
Question: `#locals` and hidden variables

### ISSUE THREAD

**Issue by andrykonchin** [ruby/prism#1776]
> @enebo this is fixed in the most recent version of prism, def foo(...); end adds *, &, and ....
>
>Overall these locals are there so that the parser can know when bar(*) is allowed (if you're within a context where * is in the local table or not). So either way, I need these in the C API. I might be okay with hiding them from the Java/Ruby API, but I do think it's nice to be able to tell in some methods where you can use * or &.

https://github.com/ruby/prism/issues/1737#issuecomment-1779575101

After the discussion in https://github.com/ruby/prism/issues/1737 there are some issues left that aren't clear to me.

1) if hidden `*`, `&` etc variables are needed for forwarding only - it makes sense for methods declaration and doesn't for blocks/lambda (because forwarding doesn't work within Proc), but `BlockNode#locals` still contain the hidden variables

```
bin/parse -e 'proc { |*| }'
@ ProgramNode (location: (1,0)-(1,12))
├── locals: []
└── statements:
    @ StatementsNode (location: (1,0)-(1,12))
    └── body: (length: 1)
        └── @ CallNode (location: (1,0)-(1,12))
            ├── receiver: ∅
            ├── call_operator_loc: ∅
            ├── message_loc: (1,0)-(1,4) = "proc"
            ├── opening_loc: ∅
            ├── arguments: ∅
            ├── closing_loc: ∅
            ├── block:
            │   @ BlockNode (location: (1,5)-(1,12))
            │   ├── locals: [:*]
            │   ├── parameters:
            │   │   @ BlockParametersNode (location: (1,7)-(1,10))
            │   │   ├── parameters:
            │   │   │   @ ParametersNode (location: (1,8)-(1,9))
            │   │   │   ├── requireds: (length: 0)
            │   │   │   ├── optionals: (length: 0)
            │   │   │   ├── rest:
            │   │   │   │   @ RestParameterNode (location: (1,8)-(1,9))
            │   │   │   │   ├── name: ∅
            │   │   │   │   ├── name_loc: ∅
            │   │   │   │   └── operator_loc: (1,8)-(1,9) = "*"
            │   │   │   ├── posts: (length: 0)
            │   │   │   ├── keywords: (length: 0)
            │   │   │   ├── keyword_rest: ∅
            │   │   │   └── block: ∅
            │   │   ├── locals: (length: 0)
            │   │   ├── opening_loc: (1,7)-(1,8) = "|"
            │   │   └── closing_loc: (1,9)-(1,10) = "|"
            │   ├── body: ∅
            │   ├── opening_loc: (1,5)-(1,6) = "{"
            │   └── closing_loc: (1,11)-(1,12) = "}"
            ├── flags: ∅
            └── name: :proc
```

2. declaring `...` parameters leads to adding `*`, `&` and `...` hidden variables but not `**` (don't know why `...` in locals could be useful but OK). So I am wondering whether missing `**` is intentional and we can rely on this behaviour:

```
bin/parse -e 'def foo(...)end'
@ ProgramNode (location: (1,0)-(1,15))
├── locals: []
└── statements:
    @ StatementsNode (location: (1,0)-(1,15))
    └── body: (length: 1)
        └── @ DefNode (location: (1,0)-(1,15))
            ├── name: :foo
            ├── name_loc: (1,4)-(1,7) = "foo"
            ├── receiver: ∅
            ├── parameters:
            │   @ ParametersNode (location: (1,8)-(1,11))
            │   ├── requireds: (length: 0)
            │   ├── optionals: (length: 0)
            │   ├── rest: ∅
            │   ├── posts: (length: 0)
            │   ├── keywords: (length: 0)
            │   ├── keyword_rest:
            │   │   @ ForwardingParameterNode (location: (1,8)-(1,11))
            │   └── block: ∅
            ├── body: ∅
            ├── locals: [:*, :&, :"..."]
            ├── def_keyword_loc: (1,0)-(1,3) = "def"
            ├── operator_loc: ∅
            ├── lparen_loc: (1,7)-(1,8) = "("
            ├── rparen_loc: (1,11)-(1,12) = ")"
            ├── equal_loc: ∅
            └── end_keyword_loc: (1,12)-(1,15) = "end"
```

3. `*` is ambiguous when both `*` and `...` are declared and forwarding probably should rely not on `locals` only but on `ParametersNode` instead (what makes `locals` not so useful for forwarding):

```
bin/parse -e 'def foo(*, ...)end'
@ ProgramNode (location: (1,0)-(1,18))
├── locals: []
└── statements:
    @ StatementsNode (location: (1,0)-(1,18))
    └── body: (length: 1)
        └── @ DefNode (location: (1,0)-(1,18))
            ├── name: :foo
            ├── name_loc: (1,4)-(1,7) = "foo"
            ├── receiver: ∅
            ├── parameters:
            │   @ ParametersNode (location: (1,8)-(1,14))
            │   ├── requireds: (length: 0)
            │   ├── optionals: (length: 0)
            │   ├── rest:
            │   │   @ RestParameterNode (location: (1,8)-(1,9))
            │   │   ├── name: ∅
            │   │   ├── name_loc: ∅
            │   │   └── operator_loc: (1,8)-(1,9) = "*"
            │   ├── posts: (length: 0)
            │   ├── keywords: (length: 0)
            │   ├── keyword_rest:
            │   │   @ ForwardingParameterNode (location: (1,11)-(1,14))
            │   └── block: ∅
            ├── body: ∅
            ├── locals: [:*, :&, :"..."]
            ├── def_keyword_loc: (1,0)-(1,3) = "def"
            ├── operator_loc: ∅
            ├── lparen_loc: (1,7)-(1,8) = "("
            ├── rparen_loc: (1,14)-(1,15) = ")"
            ├── equal_loc: ∅
            └── end_keyword_loc: (1,15)-(1,18) = "end"
```

**Comment by enebo** [comment]
My quick takes:

For 1 I think this is just convenient because we will emit a recv in runtimes for a rest but arguably if we do not forward or can access that rest reflectively we probably should not be storing it anywhere. 

For 2 @andrykonchin I also thought when @kddnewton said `...` he actually meant `**`.

For 3 is this new syntax for Ruby 3.3?  It is invalid for 3.2.

**Comment by andrykonchin** [comment]
> For 3 is this new syntax for Ruby 3.3? It is invalid for 3.2.

Oh, you are right. My bad. I haven't checked in Ruby, only with Prism (and Prism doesn't emit any syntax error).

**Comment by eregon** [comment]
>     2. declaring `...` parameters leads to adding `*`, `&` and `...` hidden variables but not `**` (don't know why `...` in locals could be useful but OK). So I am wondering whether missing `**` is intentional and we can rely on this behaviour:

I think this is because `...` on CRuby is implemented/desugared as `*, &` + `ruby2_keywords` on the method definition.
But that's a CRuby implementation detail leaking to Prism, I think that's not good.
So I think `...` should not be "desugared" by Prism, because the desugaring is implementation-specific (e.g. we desugar to `*, **, &` on TruffleRuby). i.e. it should remain as `...` and nothing else.

**Comment by kddnewton** [comment]
Yeah I got some stuff wrong here for sure.

> 1. if hidden *, & etc variables are needed for forwarding only - it makes sense for methods declaration and doesn't for blocks/lambda (because forwarding doesn't work within Proc), but BlockNode#locals still contain the hidden variables

That was just a bug - those are no longer in the list. For some reason I had convinced myself that `def foo(...) = bar(*)` was allowed, but clearly it isn't.

> 2. declaring ... parameters leads to adding *, & and ... hidden variables but not ** (don't know why ... in locals could be useful but OK). So I am wondering whether missing ** is intentional and we can rely on this behaviour:

This was actually part of the same assumption, which was also wrong. Going forward, we're only going to put `...` into the locals table, since that's the only thing that is allowed to be forwarded.

> 3. * is ambiguous when both * and ... are declared and forwarding probably should rely not on locals only but on ParametersNode instead (what makes locals not so useful for forwarding):

This was another bug made by incorrect assumptions.

All of this should be fixed by #1800.
